### ISSUE TITLE
[Bug] MariaDB with Galera unable to re-start

### ISSUE THREAD

**Issue by samip5** [mariadb-operator/mariadb-operator#959]
**Documentation**
- [x] I acknowledge that I have read the relevant [documentation](https://github.com/mariadb-operator/mariadb-operator/tree/main/docs).

**Describe the bug**

Galera is not able to recover, the pvc's are intact.

```
Name:         mariadb-galera
Namespace:    database
Labels:       app.kubernetes.io/name=mariadb
              app.kubernetes.io/part-of=mariadb
              kustomize.toolkit.fluxcd.io/name=mariadb
              kustomize.toolkit.fluxcd.io/namespace=flux-system
Annotations:  <none>
API Version:  k8s.mariadb.com/v1alpha1
Kind:         MariaDB
Metadata:
  Creation Timestamp:  2024-10-17T14:22:23Z
  Generation:          2
  Resource Version:    25849582
  UID:                 663f4bd4-cd02-4ca7-a2f5-09a648312031
Spec:
  Galera:
    Agent:
      Graceful Shutdown Timeout:  1s
      Image:                      docker-registry3.mariadb.com/mariadb-operator/mariadb-operator:0.34.0
      Kubernetes Auth:
        Enabled:           true
      Port:                5555
    Available When Donor:  false
    Config:
      Reuse Storage Volume:  false
      Volume Claim Template:
        Access Modes:
          ReadWriteOnce
        Resources:
          Requests:
            Storage:  100Mi
    Enabled:          true
    Galera Lib Path:  /usr/lib/galera/libgalera_smm.so
    Init Container:
      Image:  docker-registry3.mariadb.com/mariadb-operator/mariadb-operator:0.34.0
    Primary:
      Automatic Failover:  true
      Pod Index:           0
    Recovery:
      Cluster Bootstrap Timeout:  10m0s
      Cluster Downscale Timeout:  5m0s
      Cluster Healthy Timeout:    30s
      Cluster Monitor Interval:   10s
      Cluster Upscale Timeout:    5m0s
      Enabled:                    true
      Min Cluster Size:           1
      Pod Recovery Timeout:       5m0s
      Pod Sync Timeout:           5m0s
    Replica Threads:              1
    Sst:                          mariabackup
  Image:                          mariadb:11.5.2
  Image Pull Policy:              IfNotPresent
  Max Scale:
    Auth:
      Admin Password Secret Key Ref:
        Generate:      false
        Key:           maxscale_password
        Name:          mariadb-secret
      Admin Username:  root
      Client Password Secret Key Ref:
        Generate:            false
        Key:                 
        Name:                
      Delete Default Admin:  true
      Metrics Password Secret Key Ref:
        Generate:  false
        Key:       
        Name:      
      Monitor Password Secret Key Ref:
        Generate:  false
        Key:       
        Name:      
      Server Password Secret Key Ref:
        Generate:  false
        Key:       
        Name:      
    Connection:
      Port:         3306
      Secret Name:  mxs-galera-conn
    Enabled:        true
    Gui Kubernetes Service:
      Type:  LoadBalancer
    Kubernetes Service:
      Type:  LoadBalancer
  Max Scale Ref:
    Name:       mariadb-galera-maxscale
    Namespace:  database
  Metrics:
    Enabled:  true
    Exporter:
      Image:  prom/mysqld-exporter:v0.15.1
      Port:   9104
    Password Secret Key Ref:
      Generate:  true
      Key:       password
      Name:      mariadb-galera-metrics-password
    Service Monitor:
    Username:  mariadb-galera-metrics
  My Cnf:      [mariadb]
bind-address=*
default_storage_engine=InnoDB
binlog_format=row
innodb_autoinc_lock_mode=2
max_allowed_packet=256M

  My Cnf Config Map Key Ref:
    Key:   my.cnf
    Name:  mariadb-galera-config
  Pod Security Context:
    Fs Group:                999
    Fs Group Change Policy:  OnRootMismatch
    Run As Group:            999
    Run As User:             999
  Port:                      3306
  Primary Service:
    Type:                      LoadBalancer
  Replicas:                    3
  Replicas Allow Even Number:  false
  Resources:
    Limits:
      Memory:  2Gi
    Requests:
      Cpu:              100m
      Memory:           512Mi
  Root Empty Password:  false
  Root Password Secret Key Ref:
    Generate:  false
    Key:       password
    Name:      mariadb-secret
  Secondary Service:
    Type:  LoadBalancer
  Service:
    Type:                LoadBalancer
  Service Account Name:  mariadb-galera
  Storage:
    Ephemeral:              false
    Resize In Use Volumes:  true
    Size:                   5Gi
    Storage Class Name:     ceph-rbd
    Volume Claim Template:
      Access Modes:
        ReadWriteOnce
      Resources:
        Requests:
          Storage:           5Gi
      Storage Class Name:    ceph-rbd
    Wait For Volume Resize:  true
  Suspend:                   false
  Update Strategy:
    Type:  ReplicasFirstPrimaryLast
Status:
  Conditions:
    Last Transition Time:     2024-10-24T17:19:45Z
    Message:                  Error reconciling Galera: error recovering cluster: error getting bootstrap source: recovery status not completed
    Reason:                   Failed
    Status:                   False
    Type:                     Ready
    Last Transition Time:     2024-10-17T14:22:47Z
    Message:                  Galera initialized
    Reason:                   GaleraInitialized
    Status:                   True
    Type:                     GaleraInitialized
    Last Transition Time:     2024-10-17T14:22:48Z
    Message:                  Updated
    Reason:                   Updated
    Status:                   True
    Type:                     Updated
    Last Transition Time:     2024-10-24T17:20:17Z
    Message:                  Galera not ready
    Reason:                   GaleraNotReady
    Status:                   False
    Type:                     GaleraReady
    Last Transition Time:     2024-10-17T14:24:56Z
    Message:                  Galera configured
    Reason:                   GaleraConfigured
    Status:                   True
    Type:                     GaleraConfigured
  Current Primary:            mariadb-galera-0
  Current Primary Pod Index:  0
  Galera Recovery:
    Recovered:
      mariadb-galera-0:
        Seqno:  0
        Uuid:   00000000-0000-0000-0000-000000000000
      mariadb-galera-1:
        Seqno:  1
        Uuid:   00000000-0000-0000-0000-000000000000
      mariadb-galera-2:
        Seqno:  1
        Uuid:   00000000-0000-0000-0000-000000000000
    State:
      mariadb-galera-0:
        Safe To Bootstrap:  false
        Seqno:              -1
        Uuid:               4c5b4fb0-8c93-11ef-aa82-9257da3fe291
        Version:            2.1
      mariadb-galera-1:
        Safe To Bootstrap:  false
        Seqno:              -1
        Uuid:               4c5b4fb0-8c93-11ef-aa82-9257da3fe291
        Version:            2.1
      mariadb-galera-2:
        Safe To Bootstrap:  false
        Seqno:              -1
        Uuid:               4c5b4fb0-8c93-11ef-aa82-9257da3fe291
        Version:            2.1
Events:                     <none>
```

**Expected behaviour**

It would be able to recover.

**Steps to reproduce the bug**

1. Deploy cluster with 3 replicas using rook
2. Underlying cluster is OOM'ing.
3. It's unable to recover

**Debug information**
- Related object events:
```bash
kubectl get events --field-selector involvedObject.name=mariadb-galera -n database
LAST SEEN   TYPE     REASON             OBJECT                               MESSAGE
6m58s       Normal   SuccessfulDelete   statefulset/mariadb-galera           delete Pod mariadb-galera-2 in StatefulSet mariadb-galera successful
99s         Normal   NoPods             poddisruptionbudget/mariadb-galera   No matching pods found
119s        Normal   SuccessfulCreate   statefulset/mariadb-galera           create Pod mariadb-galera-0 in StatefulSet mariadb-galera successful
16m         Normal   SuccessfulCreate   statefulset/mariadb-galera           create Pod mariadb-galera-1 in StatefulSet mariadb-galera successful
```
- `mariadb-operator` logs: https://gist.githubusercontent.com/samip5/8171ad53035a2a03c7ac66fd14f14654/raw/456fab0c2fa62a3140fdc6480c941213c7290439/operator.log

**Environment details**:
- Kubernetes version: 1.30.3
- Kubernetes distribution: Talos v1.8.0
- mariadb-operator version: 0.34.0
- Galera data-plane version (init and agent): 0.34.0
- Install method: Helm via FluxCD2

**Additional context**

The pods seems to just over and over get re-created by the operator.

**Comment by samip5** [comment]
Tried without maxscale as well, but nope.

```
2024-10-24 20:39:55+00:00 [Note] [Entrypoint]: Entrypoint script for MariaDB Server 1:11.5.2+maria~ubu2404 started.
2024-10-24 20:39:55+00:00 [Note] [Entrypoint]: MariaDB upgrade information missing, assuming required
2024-10-24 20:39:55+00:00 [Note] [Entrypoint]: MariaDB upgrade (mariadb-upgrade or creating healthcheck users) required, but skipped due to $MARIADB_AUTO_UPGRADE setting
2024-10-24 20:39:55 0 [Note] slave_connections_needed_for_purge changed to 0 because of Galera. Change it to 1 or higher if this Galera node is also Master in a normal replication setup
2024-10-24 20:39:55 0 [Note] Starting MariaDB 11.5.2-MariaDB-ubu2404 source revision ea75a0b6009b0251e83193cd38c3974e36b65e06 server_uid WyMhKH+XvFL9LrmYaD07T6mr9pU= as process 1
2024-10-24 20:39:55 0 [Note] WSREP: Loading provider /usr/lib/galera/libgalera_smm.so initial position: 00000000-0000-0000-0000-000000000000:-1
2024-10-24 20:39:55 0 [Note] WSREP: wsrep_load(): loading provider library '/usr/lib/galera/libgalera_smm.so'
2024-10-24 20:39:55 0 [Note] WSREP: wsrep_load(): Galera 26.4.19(r5db72dad) by Codership Oy <info@codership.com> loaded successfully.
2024-10-24 20:39:55 0 [Note] WSREP: Initializing allowlist service v1
2024-10-24 20:39:55 0 [Note] WSREP: Initializing event service v1
2024-10-24 20:39:55 0 [Note] WSREP: Resolved symbol 'wsrep_node_isolation_mode_set_v1'
2024-10-24 20:39:55 0 [Note] WSREP: CRC-32C: using 64-bit x86 acceleration.
2024-10-24 20:39:55 0 [Note] WSREP: Found saved state: 00000000-0000-0000-0000-000000000000:-1, safe_to_bootstrap: 0
2024-10-24 20:39:55 0 [Note] WSREP: GCache DEBUG: opened preamble:
Version: 2
UUID: 00000000-0000-0000-0000-000000000000
Seqno: -1 - -1
Offset: -1
Synced: 1
2024-10-24 20:39:55 0 [Note] WSREP: Skipped GCache ring buffer recovery: could not determine history UUID.
2024-10-24 20:39:55 0 [Note] WSREP: Passing config to GCS: base_dir = /var/lib/mysql/; base_host = 10.244.1.115; base_port = 4567; cert.log_conflicts = no; cert.optimistic_pa = yes; debug = no; evs.auto_evict = 0; evs.delay_margin = PT1S; evs.delayed_keep_period = PT30S; evs.inactive_check_period = PT0.5S; evs.inactive_timeout = PT15S; evs.join_retrans_period = PT1S; evs.max_install_timeouts = 3; evs.send_window = 4; evs.stats_report_period = PT1M; evs.suspect_timeout = PT5S; evs.user_send_window = 2; evs.view_forget_timeout = PT24H; gcache.dir = /var/lib/mysql/; gcache.keep_pages_size = 0; gcache.keep_plaintext_size = 128M; gcache.mem_size = 0; gcache.name = galera.cache; gcache.page_size = 128M; gcache.recover = yes; gcache.size = 128M; gcomm.thread_prio = ; gcs.fc_debug = 0; gcs.fc_factor = 1.0; gcs.fc_limit = 16; gcs.fc_master_slave = no; gcs.fc_single_primary = no; gcs.max_packet_size = 64500; gcs.max_throttle = 0.25; gcs.recv_q_hard_limit = 9223372036854775807; gcs.recv_q_soft_limit = 0.25; gcs.sync_donor = no; gmcast.listen_addr 
2024-10-24 20:39:55 0 [Note] WSREP: Start replication
2024-10-24 20:39:55 0 [Note] WSREP: Connecting with bootstrap option: 0
2024-10-24 20:39:55 0 [Note] WSREP: Setting GCS initial position to 00000000-0000-0000-0000-000000000000:-1
2024-10-24 20:39:55 0 [Note] WSREP: protonet asio version 0
2024-10-24 20:39:55 0 [Note] WSREP: Using CRC-32C for message checksums.
2024-10-24 20:39:55 0 [Note] WSREP: backend: asio
2024-10-24 20:39:55 0 [Note] WSREP: gcomm thread scheduling priority set to other:0 
2024-10-24 20:39:55 0 [Note] WSREP: access file(/var/lib/mysql//gvwstate.dat) failed(No such file or directory)
2024-10-24 20:39:55 0 [Note] WSREP: restore pc from disk failed
2024-10-24 20:39:55 0 [Note] WSREP: GMCast version 0
2024-10-24 20:39:55 0 [Warning] WSREP: Failed to resolve tcp://mariadb-galera-0.mariadb-galera-internal.database.svc.cluster.local:4567
2024-10-24 20:39:55 0 [Note] WSREP: (20de4d02-aeb5, 'tcp://0.0.0.0:4567') listening at tcp://0.0.0.0:4567
2024-10-24 20:39:55 0 [Note] WSREP: (20de4d02-aeb5, 'tcp://0.0.0.0:4567') multicast: , ttl: 1
2024-10-24 20:39:55 0 [Note] WSREP: EVS version 1
2024-10-24 20:39:55 0 [Note] WSREP: gcomm: connecting to group 'mariadb-operator', peer 'mariadb-galera-0.mariadb-galera-internal.database.svc.cluster.local:,mariadb-galera-1.mariadb-galera-internal.database.svc.cluster.local:,mariadb-galera-2.mariadb-galera-internal.database.svc.cluster.local:'
2024-10-24 20:39:55 0 [Note] WSREP: (20de4d02-aeb5, 'tcp://0.0.0.0:4567') Found matching local endpoint for a connection, blacklisting address tcp://10.244.1.115:4567
2024-10-24 20:39:55 0 [Note] WSREP: (20de4d02-aeb5, 'tcp://0.0.0.0:4567') connection established to 20607561-bdaa tcp://10.244.3.171:4567
2024-10-24 20:39:56 0 [Note] WSREP: (20de4d02-aeb5, 'tcp://0.0.0.0:4567') connection established to 2121679d-b5f7 tcp://10.244.2.149:4567
2024-10-24 20:39:56 0 [Note] WSREP: EVS version upgrade 0 -> 1
2024-10-24 20:39:56 0 [Note] WSREP: declaring 20607561-bdaa at tcp://10.244.3.171:4567 stable
2024-10-24 20:39:56 0 [Note] WSREP: PC protocol upgrade 0 -> 1
2024-10-24 20:39:56 0 [Warning] WSREP: no nodes coming from prim view, prim not possible
2024-10-24 20:39:56 0 [Note] WSREP: view(view_id(NON_PRIM,20607561-bdaa,2) memb {
        20607561-bdaa,0
        20de4d02-aeb5,0
} joined {
} left {
} partitioned {
})
2024-10-24 20:39:56 0 [Note] WSREP: declaring 20607561-bdaa at tcp://10.244.3.171:4567 stable
2024-10-24 20:39:56 0 [Note] WSREP: declaring 2121679d-b5f7 at tcp://10.244.2.149:4567 stable
2024-10-24 20:39:56 0 [Warning] WSREP: no nodes coming from prim view, prim not possible
2024-10-24 20:39:56 0 [Note] WSREP: view(view_id(NON_PRIM,20607561-bdaa,3) memb {
        20607561-bdaa,0
        20de4d02-aeb5,0
        2121679d-b5f7,0
} joined {
} left {
} partitioned {
})
2024-10-24 20:39:59 0 [Note] WSREP: (20de4d02-aeb5, 'tcp://0.0.0.0:4567') turning message relay requesting off
2024-10-24 20:40:26 0 [ERROR] WSREP: failed to open gcomm backend connection: 110: failed to reach primary view: 110 (Connection timed out)
         at ./gcomm/src/pc.cpp:connect():160
2024-10-24 20:40:26 0 [ERROR] WSREP: ./gcs/src/gcs_core.cpp:gcs_core_open():221: Failed to open backend connection: -110 (Connection timed out)
2024-10-24 20:40:27 0 [ERROR] WSREP: ./gcs/src/gcs.cpp:gcs_open():1681: Failed to open channel 'mariadb-operator' at 'gcomm://mariadb-galera-0.mariadb-galera-internal.database.svc.cluster.local,mariadb-galera-1.mariadb-galera-internal.database.svc.cluster.local,mariadb-galera-2.mariadb-galera-internal.database.svc.cluster.local': -110 (Connection timed out)
2024-10-24 20:40:27 0 [ERROR] WSREP: gcs connect failed: Connection timed out
2024-10-24 20:40:27 0 [ERROR] WSREP: wsrep::connect(gcomm://mariadb-galera-0.mariadb-galera-internal.database.svc.cluster.local,mariadb-galera-1.mariadb-galera-internal.database.svc.cluster.local,mariadb-galera-2.mariadb-galera-internal.database.svc.cluster.local) failed: 7
2024-10-24 20:40:27 0 [ERROR] Aborting
```

**Comment by mmontes11** [comment]
Hey there @samip5 ! Thanks for reporting this and the level of detail, it has been very useful

Currently, this state is not recoverable, as there is a `Pod` with `0` sequence number, something that makes the `isComplete` function return false:
```bash
  Galera Recovery:
    Recovered:
      mariadb-galera-0:
        Seqno:  0
        Uuid:   00000000-0000-0000-0000-000000000000
      mariadb-galera-1:
        Seqno:  1
        Uuid:   00000000-0000-0000-0000-000000000000
      mariadb-galera-2:
        Seqno:  1
        Uuid:   00000000-0000-0000-0000-000000000000
 ```
See the `isComplete` function for reference:
https://github.com/mariadb-operator/mariadb-operator/blob/85d8b62e913ba33a9f5bfcff9aaaada04f2bf5b9/pkg/controller/galera/recovery_status.go#L145

This could be a corner case we haven't covered, as it might be possible to bootstrap from either `mariadb-galera-1` and `mariadb-galera-2` even if `mariadb-galera-0` has sequence `0`. 

Luckly enough, we have a way to verify this, we could bypass the current recovery process by using `forceClusterBootstrapInPod`. See the docs:
- https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/GALERA.md#force-cluster-bootstrap

Please try the following steps:
- Set `forceClusterBootstrapInPod=mariadb-galera-0`. Since you are using flux, I would commit this changes and let flux eventually reconcile:
```yaml
apiVersion: k8s.mariadb.com/v1alpha1
kind: MariaDB
metadata:
  name: mariadb-galera
spec:
  ...
  galera:
    enabled: true
    recovery:
      enabled: true
      forceClusterBootstrapInPod: "mariadb-galera-0"
``` 
- Once the `forceClusterBootstrapInPod` is in place, restart the operator to reset the exponential backoff and trigger the recovery process
- If successful, remember to revert `forceClusterBootstrapInPod`

As an alternative, remember that you can always bootstrap an instance from a backup:
- https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/BACKUP.md#bootstrap-new-mariadb-instances

Let me know how it goes!

**Comment by mmontes11** [comment]
> Tried without maxscale as well, but nope.

By the way, MaxScale should make no difference in the Galera cluster recovery

**Comment by mmontes11** [comment]
@samip5 I'll be releasing a `0.35.1` patch version with this fix shortly, stay tuned!

**Comment by samip5** [comment]
That did not fix it.

```
Name:         mariadb-galera
Namespace:    database
Labels:       app.kubernetes.io/name=mariadb
              app.kubernetes.io/part-of=mariadb
              kustomize.toolkit.fluxcd.io/name=mariadb
              kustomize.toolkit.fluxcd.io/namespace=flux-system
Annotations:  <none>
API Version:  k8s.mariadb.com/v1alpha1
Kind:         MariaDB
Metadata:
  Creation Timestamp:  2024-10-24T20:24:12Z
  Generation:          4
  Resource Version:    27865410
  UID:                 21806b04-007c-446c-b727-7125b2f50d32
Spec:
  Galera:
    Agent:
      Graceful Shutdown Timeout:  1s
      Image:                      docker-registry3.mariadb.com/mariadb-operator/mariadb-operator:0.34.0
      Kubernetes Auth:
        Enabled:           true
      Port:                5555
    Available When Donor:  false
    Config:
      Reuse Storage Volume:  false
      Volume Claim Template:
        Access Modes:
          ReadWriteOnce
        Resources:
          Requests:
            Storage:  100Mi
    Enabled:          true
    Galera Lib Path:  /usr/lib/galera/libgalera_smm.so
    Init Container:
      Image:  docker-registry3.mariadb.com/mariadb-operator/mariadb-operator:0.34.0
    Primary:
      Automatic Failover:  true
      Pod Index:           0
    Recovery:
      Enabled:               false
      Pod Recovery Timeout:  15m0s
    Replica Threads:         1
    Sst:                     mariabackup
  Image:                     mariadb:11.5.2
  Image Pull Policy:         IfNotPresent
  Metrics:
    Enabled:  true
    Exporter:
      Image:  prom/mysqld-exporter:v0.15.1
      Port:   9104
    Password Secret Key Ref:
      Generate:  true
      Key:       password
      Name:      mariadb-galera-metrics-password
    Service Monitor:
    Username:  mariadb-galera-metrics
  My Cnf:      [mariadb]
bind-address=*
default_storage_engine=InnoDB
binlog_format=row
innodb_autoinc_lock_mode=2
max_allowed_packet=256M

  My Cnf Config Map Key Ref:
    Key:   my.cnf
    Name:  mariadb-galera-config
  Pod Security Context:
    Fs Group:                999
    Fs Group Change Policy:  OnRootMismatch
    Run As Group:            999
    Run As User:             999
  Port:                      3306
  Primary Service:
    Type:                      LoadBalancer
  Replicas:                    3
  Replicas Allow Even Number:  false
  Resources:
    Limits:
      Memory:  2Gi
    Requests:
      Cpu:              100m
      Memory:           512Mi
  Root Empty Password:  false
  Root Password Secret Key Ref:
    Generate:  false
    Key:       password
    Name:      mariadb-secret
  Secondary Service:
    Type:  LoadBalancer
  Service:
    Type:                LoadBalancer
  Service Account Name:  mariadb-galera
  Storage:
    Ephemeral:              false
    Resize In Use Volumes:  true
    Size:                   5Gi
    Storage Class Name:     ceph-rbd
    Volume Claim Template:
      Access Modes:
        ReadWriteOnce
      Resources:
        Requests:
          Storage:           5Gi
      Storage Class Name:    ceph-rbd
    Wait For Volume Resize:  true
  Suspend:                   false
  Update Strategy:
    Type:  ReplicasFirstPrimaryLast
Status:
  Conditions:
    Last Transition Time:  2024-10-24T20:24:12Z
    Message:               Error reconciling Galera: error recovering cluster: error performing recovery: 1 error occurred:
                           * error building recovery Job for Pod 'mariadb-galera-2': Galera recovery must be enabled


    Reason:                   Failed
    Status:                   False
    Type:                     Ready
    Last Transition Time:     2024-10-24T20:24:12Z
    Message:                  Galera configured
    Reason:                   GaleraConfigured
    Status:                   True
    Type:                     GaleraConfigured
    Last Transition Time:     2024-10-24T20:24:12Z
    Message:                  Galera not ready
    Reason:                   GaleraNotReady
    Status:                   False
    Type:                     GaleraReady
  Current Primary:            mariadb-galera-0
  Current Primary Pod Index:  0
  Galera Recovery:
    State:
      mariadb-galera-1:
        Safe To Bootstrap:  false
        Seqno:              -1
        Uuid:               00000000-0000-0000-0000-000000000000
        Version:            2.1
      mariadb-galera-2:
        Safe To Bootstrap:  false
        Seqno:              -1
        Uuid:               00000000-0000-0000-0000-000000000000
        Version:            2.1
Events:                     <none>
```

**Comment by samip5** [comment]
It seems that despite having "spec.recovery" things, without enabled: true, it wouldn't actually be enabled.

**Comment by mmontes11** [comment]
>  It seems that despite having "spec.recovery" things, without enabled: true, it wouldn't actually be enabled.

Very weird, can't really see how this could end up being `false`, unless there something very wrong I am missing either in the operator or on the controlplane side:
- https://github.com/mariadb-operator/mariadb-operator/blob/b3365c13ec3e0f644fddea833fec7d19dce677e4/pkg/builder/batch_builder.go#L346

Please try with the following steps
- Double check that the operator is running with the 0.35.1. Follow the [upgrade guide ](https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/releases/UPGRADE_0.35.1.md)if it's not.
- Manually delete your `MariaDB` resource. This will leave your PVCs untouched
- Let flux recreate the `MariaDB` 
- When created, a recovery process will be automatically triggered, since the PVCs are still there.

**Comment by samip5** [comment]
> * Manually delete your `MariaDB` resource. This will leave your PVCs untouched

I don't believe that's true if the underlying storage has Delete as the retainPolicy on the storageclass.


I was able to recover, but only after setting:

```
  galera:
    enabled: true
    recovery:
      enabled: true
      forceClusterBootstrapInPod: "mariadb-galera-1"
```


So currently the cluster is working again.

**Comment by mmontes11** [comment]
> I don't believe that's true if the underlying storage has Delete as the retainPolicy on the storageclass.

The `retainPolicy` of the PVCs would apply when the PVCs are deleted. If you take a closer look, our PVCs don't have `ownerReferences` (unlike other operators):

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    pv.kubernetes.io/bind-completed: "yes"
    pv.kubernetes.io/bound-by-controller: "yes"
    volume.beta.kubernetes.io/storage-provisioner: storage.rbd.csi.ceph.com
    volume.kubernetes.io/storage-provisioner: storage.rbd.csi.ceph.com
  labels:
    app.kubernetes.io/instance: mariadb-xiaowen
    app.kubernetes.io/name: mariadb
    pvc.k8s.mariadb.com/role: storage
  name: storage-mariadb-xiaowen-0
  namespace: media
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  storageClassName: rook-ceph
  volumeName: pvc-49218cf9-4103-4974-bc86-9c2ba1362793
```
Therefore, when the parent resource i.e. `MariaDB` and then the `StatefulSet` are deleted, no delete operation is performed on them, so it should be fine.

> I was able to recover

Glad to hear this! Let me know if you manage to reproduce the previous error consistently, please do open another issue if this problem re-appears:

> Error reconciling Galera: error recovering cluster: error performing recovery: 1 error occurred:
                           * error building recovery Job for Pod 'mariadb-galera-2': Galera recovery must be enabled
